pkgs/development/cuda-modules/_cuda/lib/db.nix

```diff
@@ -0,0 +1,200 @@
+        );
+      predicates = {
+        "<" = lib.versionOlder;
+        ">=" = lib.versionAtLeast;
+      };
+
+      # ∷ PName ⇒ VersionString
+      pinnedVersions =
+        lib.concatMapAttrs (
+          productName: productVersion:
+          maybeOr (selectQuantized db.release.package.${productName} productVersion) { }
+        ) pinProducts
+        // lib.mapAttrs (
+          pname: version:
+          lib.last (
+            lib.attrNames (
+              lib.filterAttrs (versionKey: _: matchQuantized version versionKey) db.package.version.${pname}
+            )
+          )
+        ) pinPackages;
+
+      # ∷ PName → { version: VersionString, blobs: SHA256 ⇒ { URL, SystemStringNvidia... } }
+      latestSupported' =
+        pname:
+        lib.last (
+          [ null ]
+          ++ lib.mapAttrsToList (version: blobs: { inherit version blobs; }) supportedArchives.${pname}
+        );
+
+      # ∷ PName ⇒ { version: VersionString, blobs: SHA256 ⇒ { URL, SystemStringNvidia... } }
+      latestSupported = lib.mapAttrs (pname: _: latestSupported' pname) db.package.pname;
+
+      # ∷ PName ⇒ VersionString ⇒ SHA256 ⇒ { URL, SystemStringNvidia, ... }
+      supportedArchives = lib.mapAttrs (
+        pname: _:
+        lib.addErrorContext "while evaluating `supportedArchives' ... ${pname}`" (
+          supportedArchives' db.package.version pname
+        )
+      ) db.package.version;
+
+      # ∷ (PName ⇒ VersionString ⇒ SHA256 ⇒ ()) → PName ⇒ VersionString ⇒ SHA256 ⇒ { URL, SystemStringNvidia, ... }
+      supportedArchives' =
+        knownVersions: pname:
+        lib.concatMapAttrs (
+          version: hashes:
+          let
+            supportedBlobs = lib.concatMapAttrs (
+              sha256: _:
+              let
+                supported = archiveIsSupported archive;
+                archive = db.archive.sha256.${sha256};
+              in
+              lib.optionalAttrs supported { ${sha256} = archive; }
+            ) hashes;
+          in
+          lib.optionalAttrs (supportedBlobs != { }) { ${version} = supportedBlobs; }
+        ) knownVersions.${pname};
+
+      reduceConstraints = lib.mapNullable (
+        archives:
+        builtins.head (
+          lib.mapAttrsToList (
+            sha256: archive:
+            assert (
+              lib.assertMsg (testAllConstraints archive.extraConstraints) "while testing constraints of ${sha256}"
+            );
+            lib.removeAttrs archive [ "extraConstraints" ] // { inherit sha256; }
+          ) archives
+        )
+      );
+
+      solution =
+        # Versioned attributes (e.g. `cudnn_8_9`) kept for compatibility
+        lib.flip lib.concatMapAttrs (lib.removeAttrs db.release.package [ "cuda" ]) (
+          productName: productVersions:
+          lib.concatMapAttrs (
+            _productVersion: packageVersions:
+            lib.mapAttrs' (
+              pname: version:
+              lib.nameValuePair (mkVersionedPackageName productName version) {
```

Comment:
Why does `mkVersionedPackageName` use `version` and not `productVersion`?

Reply:
Why not use the more specific version?